City-of-Helsinki/hel-fi-drupal-grants
# feat: AU-2329, E2E: Test tooltip functionality

# [AU-2541](https://helsinkisolutionoffice.atlassian.net/browse/AU-2541)

## What was done

This pull request implements the possibility to perform `tooltip` validation on application forms. As an example, tooltip validation has been added to the `draft` variant of application `64`.

## What has changed

- Tooltip validation can now be added to any form variant by setting the `validateTooltips` key to `true`.

- **Configuration**:

  - You must specify a `tooltipsToValidate` key for each form page you wish to validate tooltips on.
  
  - The `tooltipsToValidate` key accepts a `TooltipsList`, which is an array of the type `Tooltip`.

- **Tooltip definition**

  - Each tooltip should define an aria-label under the `aria_label` key that triggers the tooltip when clicked.
  
  - Each tooltip should also define a tooltip message under the `message` key.

![Screenshot 2024-06-19 at 11 17 53](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/591fe36d-a0c4-479d-9404-ea93b100f748)

## Other smaller changes

- Refactored code related to navigation into `navigation_helpers.ts`.

- Fixed a few comments.

## How to install

Make sure your instance is up and running on correct branch.
* `git checkout feature/AU-2329-e2e-test-js-functionality`
* `make fresh`
* `make drush-cr`

## How to test

- Remove `ENABLED_FORM_VARIANTS` and `DISABLED_FORM_VARIANTS` values in the `.env` file.

- Run `make test-pw-p PROJECT=forms-64`. Make sure that all the tests pass, and that you see this during the tooltip validation test:

![Screenshot 2024-06-19 at 14 40 51](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/39242afd-d39d-4a52-8f9c-e4f0c84ceee2)


[AU-2541]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2541?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
